### ISSUE TITLE
HCL parsing does not work with modules that have a `source` in a private Terraform registry

### ISSUE THREAD

**Issue by aliscott** [infracost/infracost#1667]
HCL parsing does not work if the Terraform code contains a module with a source in a private Terraform registry. Private git modules work ok.

Example this code:
```hcl
provider "aws" {
  region                      = "us-east-1"
  skip_credentials_validation = true
  skip_requesting_account_id  = true
  access_key                  = "mock_access_key"
  secret_key                  = "mock_secret_key"
}

module "ec2_cluster" {
  source  = "app.terraform.io/infracost/ec2-instance/aws"
  version = "~> 2.0"

  name           = "my-cluster"
  instance_count = 5

  ami                    = "ami-ebd02392"
  instance_type          = "t2.micro"
  key_name               = "user1"
  monitoring             = true
  vpc_security_group_ids = ["sg-12345678"]
  subnet_id              = "subnet-eddcdzz4"
}
```

Shows this error:
```
Error: Error loading Terraform modules: error downloading 'file:///examples/terraform/.infracost/terraform_modules/ec2_cluster/app.terraform.io/infracost/ec2-instance/aws': source path error: stat /examples/terraform/.infracost/terraform_modules/ec2_cluster/app.terraform.io/infracost/ec2-instance/aws: no such file or directory
```

There's two issues which are causing this:
1. Our code assumes that the module API is always at `<host>/v1/modules` but this can be different for different registries. We should use the [terraform-svchost](https://github.com/hashicorp/terraform-svchost/tree/master/disco) package to discover the correct service URLs
2. Our code does not pass through any credentials when pulling remote registry modules.

I've created a [work in progress branch here](https://github.com/infracost/infracost/compare/fix/hcl-private-registry-modules?expand=1).

**Comment by EppO** [comment]
if you want feedback about this feature, I can test your changes with a GitLab Terraform registry which uses "<host>/terraform/<project name>/azurerm" module address

**Comment by alikhajeh1** [comment]
@EppO thanks! Just to double check, you're using a private TF module hosted on [GitLab Terraform Registry](https://docs.gitlab.com/ee/user/packages/terraform_module_registry/)? I've not used that before, does GitLab support public modules too? Or would those just be public git repos?

**Comment by EppO** [comment]
I guess that depends if your repo is public or private. I would assume it supports both

**Comment by EppO** [comment]
just tried, even though the repo is public (you can browse the repo through the UI and even download the module tarball yourself), you can't terraform init to download the module until you have your gitlab token set up. Good news is that a `$CI_JOB_TOKEN` is enough to download it.

**Comment by alikhajeh1** [comment]
@EppO great - that's helpful to know.

**Comment by hugorut** [comment]
Private Terraform registry support has been merged into master and will be available as part of the `0.10.7` release. I'll ping here when it's out. @EppO if you have any time I'd love to take you up on your offer to test Gitlab registries. Is there any chance that I can get you to try out master. You'll need to:

1. have Go installed on your machine (1.18)
2. clone the Infracost repo
3. run `make build` in the root of the repo
4. then use the binary built under the `build` folder on your project, e.g. `/path/to/cloned/repo/build/infracost breakdown --path my-project`

If that's a little much I can ping you when `0.10.7` is out and you can test with an updated CLI 🙏

**Comment by EppO** [comment]
I did a bit of troubleshooting, the download URL looks good: 

```bash
curl -v --header "Authorization: Bearer ${INFRACOST_TERRAFORM_CLOUD_TOKEN}" "https://<mygitlabhostname>/api/v4/packages/terraform/modules/v1/terraform/my_module/azurerm/1.12.1/download"

*   Trying x.x.x.x:443...
* Connected to <mygitlabhostname> (x.x.x.x) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
* (304) (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
... snip cert data ...
*  SSL certificate verify ok.
> GET /api/v4/packages/terraform/modules/v1/terraform/my_module/azurerm/1.12.1/download HTTP/1.1
> Host: <mygitlabhostname>
> User-Agent: curl/7.79.1
> Accept: */*
> Authorization: Bearer <token value>
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 204 No Content
< Server: nginx
< Date: Wed, 29 Jun 2022 17:52:30 GMT
< Connection: keep-alive
< Cache-Control: no-cache
< Vary: Origin
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-Request-Id: 01G6R9FNYTJQ1W7FDDWNZNRSWH
< X-Runtime: 0.058537
< X-Terraform-Get: /api/v4/packages/terraform/modules/v1/terraform/my_module/azurerm/1.12.1/file?token=<a very very very long token>&archive=tgz
< Strict-Transport-Security: max-age=63072000
< Referrer-Policy: strict-origin-when-cross-origin
<
* Connection #0 to host <mygitlabhostname> left intact
* 
```

If I use the `X-Terraform-Get` value and make another curl request, I can download the tgz:

```bash
curl -v --output module.tgz  "https://<mygitlabhostname>/api/v4/packages/terraform/modules/v1/terraform/my_module/azurerm/1.12.1/file?token=<the same very very very long token from X-Terraform-Get>&archive=tgz"
* Connected to <mygitlabhostname> (x.x.x.x) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (OUT), TLS handshake, Client hello (1):
} [338 bytes data]
* (304) (IN), TLS handshake, Server hello (2):
{ [87 bytes data]
* TLSv1.2 (IN), TLS handshake, Certificate (11):
{ [2988 bytes data]
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
{ [333 bytes data]
* TLSv1.2 (IN), TLS handshake, Server finished (14):
{ [4 bytes data]
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
} [70 bytes data]
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.2 (OUT), TLS handshake, Finished (20):
} [16 bytes data]
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
{ [1 bytes data]
* TLSv1.2 (IN), TLS handshake, Finished (20):
{ [16 bytes data]
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
... snip cert data ...
*  SSL certificate verify ok.
> GET /api/v4/packages/terraform/modules/v1/terraform/my_module/azurerm/1.12.1/file?token=<a very very very long token>&archive=tgz HTTP/1.1
> Host: <mygitlabhostname>
> User-Agent: curl/7.79.1
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: nginx
< Date: Wed, 29 Jun 2022 17:59:44 GMT
< Content-Type: application/octet-stream
< Content-Length: 1625
< Connection: keep-alive
< Accept-Ranges: bytes
< Cache-Control: max-age=0, private, must-revalidate
< Content-Disposition: attachment; filename="my_module-azurerm-1.12.1.tgz"; filename*=UTF-8''my_module-azurerm-1.12.1.tgz
< Content-Transfer-Encoding: binary
< Etag: W/"12ae32cb1ec02d01eda3581b127c1fee"
< Last-Modified: Thu, 21 Apr 2022 04:52:23 GMT
< Vary: Origin
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-Request-Id: 01G6R9WXRS0HBWTYMJGP8FB50H
< X-Runtime: 0.071688
< Strict-Transport-Security: max-age=63072000
< Referrer-Policy: strict-origin-when-cross-origin
<
{ [1625 bytes data]
100  1625  100  1625    0     0   7038      0 --:--:-- --:--:-- --:--:--  7286
* Connection #0 to host <mygitlabhostname> left intact

```

module tarball looks ok:
```bash
tar xvzf module.tgz
x ./
x ./outputs.tf
x ./variables.tf
x ./provider.tf
x ./main.tf
x ./README.md
```

**Comment by EppO** [comment]
bottom line, I think it has hard time to use the `X-Terraform-Get` value from the first request, but I would need more debug around the actual request it's trying to make. I'll keep investigating

**Comment by EppO** [comment]
following up with my investigation: I don't know about Terraform official registry but GitLab doesn't return the hostname part in `X-Terraform-Get`, just the URI. Hence the error: 
```
source path error: stat /api/v4/packages/terraform/modules/v1/terraform/my_module/azurerm/1.12.1/file: no such file or directory
```

**Comment by hugorut** [comment]
@EppO thanks for all this manual debugging, yup it looks like that might be the issue right there. Normally `x-terraform-get` returns a fully qualified URL which we use to download the tar. I'll put a fix in for this tomorrow (it's late here in the UK) and ping you to test.

**Comment by EppO** [comment]
in https://www.terraform.io/registry/api-docs#download-source-code-for-a-specific-module-version, docs say:

> The value of X-Terraform-Get may instead be a relative URL, indicated by beginning with /, ./ or ../, in which case it is resolved relative to the full URL of the download endpoint.

so I assume infracost needs to support both, relative and absolute. 

> @EppO thanks for all this manual debugging, yup it looks like that might be the issue right there. Normally `x-terraform-get` returns a fully qualified URL which we use to download the tar. I'll put a fix in for this tomorrow (it's late here in the UK) and ping you to test.

no rush, let me know when I can try again.

**Comment by EppO** [comment]
go-getter supports [redirects](https://github.com/hashicorp/go-getter#user-content-security-options) via the `X-Terraform-Get` header, can't you just use go-getter with the download module URL from the get-go?

**Comment by aliscott** [comment]
Thanks @EppO - I've added a fix for the relative download paths in https://github.com/infracost/infracost/pull/1795.

**Comment by alikhajeh1** [comment]
@aliscott should we re-open this issue until gitlab support is confirmed as working by @EppO or should we create a new issue to track it? (@EppO I really appreciate your help with that)

**Comment by alikhajeh1** [comment]
bah, ignore my previous comment, I see the pull request has been merged with confirmation comments.

**Comment by hugorut** [comment]
Private module support is now available in the [0.10.7](https://github.com/infracost/infracost/releases/tag/v0.10.7) release.
